### ISSUE TITLE
Mac M1 binaries

### ISSUE THREAD

**Issue by vorburger** [MariaDB4j/MariaDB4j#497]
For MariaDB4j to "just" work on M1 Mac... either:

A. Someone contributes binaries for MariaDB binaries JAR for M1

-OR-

B. #296 is implemented to support running MariaDB in a container instead of from binaries in a JAR

I welcome community contributions for A. (but won't work on that myself), but think B. is a better way to pursue for the future.

**Comment by vorburger** [comment]
https://github.com/docker/for-mac/issues/5389 has something about (now solved?) performance issues on M1.

https://stackoverflow.com/questions/69896059/is-it-possible-to-get-mariadb4j-to-work-on-an-m1-mac asks for this.

**Comment by vorburger** [comment]
Just like for #298 re. ARM, https://downloads.mariadb.org doesn't actually seem to have DB binaries for M1, but presumably someone interested in supporting this knows of some place where prebuilt binaries for Mac M1 arch can be downloaded from.

**Comment by balamsoft** [comment]
Well, there's a really easy way to build MariaDB on Mac using Homebrew.
`brew install mariadb --build-from-source` 

As of now this command will build MariaDB 10.6.4.

**Comment by balamsoft** [comment]
I have built it using an M1 Mac Mini and I have the binaries.
```
➜  server git:(109fc67d4da) brew install mariadb --build-from-source

==> Downloading https://ghcr.io/v2/homebrew/core/groonga/manifests/11.0.9
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/groonga/blobs/sha256:891b7a85bf8e539d276f3251648c471ce925e401f99b543316ad3c39d794b801
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:891b7a85bf8e539d276f3251648c471ce925e401f99b543316ad3c39d794b801?
######################################################################## 100.0%
==> Downloading https://downloads.mariadb.com/MariaDB/mariadb-10.6.4/source/mariadb-10.6.4.tar.gz
######################################################################## 100.0%
==> Installing dependencies for mariadb: groonga
==> Installing mariadb dependency: groonga
==> Pouring groonga--11.0.9.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/groonga/11.0.9: 940 files, 51.3MB
==> Installing mariadb
==> Patching
==> cmake . -DMYSQL_DATADIR=/opt/homebrew/var/mysql -DINSTALL_INCLUDEDIR=include/mysql -DINSTALL_MANDIR=share/man -DINSTALL_DOCDIR=share/doc/mariadb -
==> make
==> make install
==> Caveats
A "/etc/my.cnf" from another install may interfere with a Homebrew-built
server starting up correctly.

MySQL is configured to only allow connections from localhost by default

To start mariadb now and restart at login:
  brew services start mariadb
Or, if you don't want/need a background service you can just run:
  mysql.server start
==> Summary
🍺  /opt/homebrew/Cellar/mariadb/10.6.4: 897 files, 172.2MB, built in 3 minutes 6 seconds
==> Running `brew cleanup mariadb`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
==> Caveats
==> mariadb
A "/etc/my.cnf" from another install may interfere with a Homebrew-built
server starting up correctly.

MySQL is configured to only allow connections from localhost by default

To start mariadb now and restart at login:
  brew services start mariadb
Or, if you don't want/need a background service you can just run:
  mysql.server start
```

**Comment by vorburger** [comment]
> brew install mariadb --build-from-source
> Would those be of any use?

The only problem with this approach is that you cannot easily contribute that as a PR to this repo and make it possible for myself or anyone sel to easily build from source accross platforms - that will only work on a Mac (which e.g. I or other others don't have / use).

I'm unfortuantely not willing (anymore) to trust 3rd-party and include native binaries built by strangers on the Internet in this Git repo; I hope you understand. (I will hower accept PRs which [DL binaries from "reputable sources" such as `downloads.mariadb.com`](https://github.com/vorburger/MariaDB4j/blob/master/DBs/mariaDB4j-db-10.3/mariaDB4j-db-winx64-10.3/prepare.xml).)

A possible "compromise" I can think of to unblock this could be to make a PR which builds M1 binaries IF RUN ON A M1 MAC, and then let Mac end-users of this project decide if they want to `mvn install | deploy` that themselves. I'm happy to review and merge a PR for that. But I won't be able to locally run it myself and deploy it to Maven Central with future release.

**Comment by balamsoft** [comment]
> I'm unfortuantely not willing (anymore) to trust 3rd-party and include native binaries built by strangers on the Internet in this Git repo; I hope you understand.
Understood.

**Comment by balamsoft** [comment]
Maybe another way to approach this would be detect if the `mariadb` binaries are already available at the host.

**Comment by balamsoft** [comment]
if anyone's interested I was able to build the latest MariaDB4j version for Mac M1 processors.
Steps:

1. Clone MariaDB4j.
2. Build MariaDB from source (`arch -arm64 brew install mariadb@10.2 --build-from-source`).
3. Go to folder `/opt/homebrew/Cellar/mariadb@10.2/10.2.41` and copy all the required binaries from the `bin` and `script` subfolders to `MariaDB4j/DBs/mariaDB4j-db-mac64-10.2.11/src/main/resources/ch/vorburger/mariadb4j/mariadb-10.2.11/osx/bin`.
4. In a terminal go to `MariaDB4j/DBs/mariaDB4j-db-mac64-10.2.11` and do a `mvn install`.

That's it, now your tests will run in your M1 :)

**Comment by vorburger** [comment]
> `brew install mariadb --build-from-source`
> Maybe another way to approach this would be detect if the mariadb binaries are already available at the host.

This may actually be the simplest way forward on this topic...

Watch new issue #560 (which is about the generalization of this idea, not Mac & M1 specific).

**Comment by vorburger** [comment]
PS: Anyone reading this and using MariaDB4j, please [Sponsor my maintenance this project](https://github.com/sponsors/vorburger)! I would very much appreciate it.

**Comment by stevenodb** [comment]
I was trying to add the AppleSilicon (arm64) binaries, but the current structure only supports win32, linux, osx; adding another architecture for osx isn't supported, for instance. It would need some refactoring to how platform detection is currently designed. Do you agree?

**Comment by vorburger** [comment]
Steven, you can just name it something like osx-arm64/ perhaps, without
needing to refactor the existing one?

BTW please note that I cannot accept PR with binaries, so if you are
willing to make a contribution for this (thank you!), then it would have to
download an archive from a reputable source and unzip it during the build,
using something similar to here:
https://github.com/vorburger/MariaDB4j/blob/master/DBs/mariaDB4j-db-10.3/mariaDB4j-db-mac64-10.3/prepare.xml

**Comment by Renevith** [comment]
The Homebrew recipe at https://formulae.brew.sh/formula/mariadb says "Bottle (binary package) installation support provided for: [...] Apple Silicon". Does anyone know if Homebrew would allow the MariaDB4j build to point to their bottle URL, and if so, what URL to use? The URLs behind the "Bottle JSON API" link e.g. [this one](https://ghcr.io/v2/homebrew/core/mariadb/blobs/sha256:26865f1cd7779f31e24637b02ab9dfb2bced57e00314f1e345eae444a6b41560) just return an "authentication required" error for me.

**Comment by vorburger** [comment]
#560 is currently the recommended approach for how to use this library on Mac M1 models.

**Comment by stevenodb** [comment]
I have a [branch with working support for macos-arm64](https://github.com/stevenodb/MariaDB4j/tree/feature/platform_support_macos-arm64) in the manner [you suggested](https://github.com/vorburger/MariaDB4j/issues/497#issuecomment-1193393759), but no eyes on a reputable source to host an apple silicon build of mariadb.

**Comment by vorburger** [comment]
> I have a [draft PR with working support for macos-arm64](https://github.com/vorburger/MariaDB4j/pull/628) in the manner [you suggested](https://github.com/vorburger/MariaDB4j/issues/497#issuecomment-1193393759),

@stevenodb thanks for #628!

> but no leads on a reputable source to host an apple silicon build of mariadb.

Yeah that's still the crux here, given that https://mariadb.org/download doesn't seem to have Mac builds. So I guess the options are:

A. Someone to find and post here a publicly accessible URL to Homebrew (or any other reputable site) from where I can (have the Ant script in #628) myself download a MariaDB version for Apple Silicon.

B. Explain to me how I could myself use `brew` on Linux to install the an Apple Silicon version of MariaDB on Linux, which I can then package. (But I'm unclear about how Homebrew exactly works and if it even support this kind of cross installation; and it sounds like a PITA.)

C. Go on vacation to a friend who has a M1 Mac who can install MariaDB and copy the files from there, and commit them into this Git repository. I don't particularly like this idea, and don't really have much time for doing that.

D. Automate C. with scripts in this repo and execute those on [a GitHub supported Runner on Mac OS](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) to produce what we want. I guess if this sounds like your idea of fun, you could go for it... this sounds like it could be a "least PITA" and "most future proof maintenable" route forward here.

E. Have the MariaDB4j code assume Mac M1 users have `brew` installed (or even automate download its installation?), and `exec`something like `brew install mariadb@10.6` (to a temporary directory?) to solve this conundrum. Possible. Contrived. Not sure how useful this is. Try, if you are so inclined.

F. Give up on this entirely? 😈  Is running `brew install mariadb@10.6` yourselves and then using https://github.com/vorburger/MariaDB4j#how-using-existing-native-mariadb-binaries really **THAT _hard_** and inconvenient? Why?

For full disclosure and expliciteness, I don't really have the time to investigate these further - but I'm happy to review "community initiaves" by readers of this post who would like to explore doing any of this.

PS: How about someone _SPONSOR_ this? 😹 E.g. [Tidelift](https://tidelift.com), [IssueHunt](https://issuehunt.io/r/vorburger/MariaDB4j/issues/497), [LFX Crowdfunding](https://crowdfunding.lfx.linuxfoundation.org/projects/MariaDB4j).

**Comment by ranma2913** [comment]
> if anyone's interested I was able to build the latest MariaDB4j version for Mac M1 processors. Steps:
> 
>     1. Clone MariaDB4j.
> 
>     2. Build MariaDB from source (`arch -arm64 brew install mariadb@10.2 --build-from-source`).
> 
>     3. Go to folder `/opt/homebrew/Cellar/mariadb@10.2/10.2.41` and copy all the required binaries from the `bin` and `script` subfolders to `MariaDB4j/DBs/mariaDB4j-db-mac64-10.2.11/src/main/resources/ch/vorburger/mariadb4j/mariadb-10.2.11/osx/bin`.
> 
>     4. In a terminal go to `MariaDB4j/DBs/mariaDB4j-db-mac64-10.2.11` and do a `mvn install`.
> 
> 
> That's it, now your tests will run in your M1 :)

Hi, I wanted to share my complete command list which should work on anyone's Apple M1/Arm64 Chip. 

### Apple M1/arm64 Chips: Build MariaDB from Source and Override Binaries in local MariaDB4j
1. Check mariadb major.minor.patch version in your error logs and update the variables in the commands.
![image](https://user-images.githubusercontent.com/4295880/209835389-3705a6ce-3c33-4f7a-9546-8b54f89bc9f2.png)
2. Update the `MariaDB4j_Workspace` variable to point to your project workspace directory.
3. Run commands for the required mariadb version:
```shell
export MariaDB4j_MAJORMINOR="10.2" && export MariaDB4j_VERSION="${MariaDB4j_MAJORMINOR}.11"
brew install mariadb@$MariaDB4j_MAJORMINOR --build-from-source
export MariaDB4j_Workspace=/Users/$(whoami)/WORK/RIPTiDE/IntelliJ_Workspace && export MariaDB4j_Repo=${MariaDB4j_Workspace}/MariaDB4j 
env | grep MariaDB4j_
git clone https://github.com/vorburger/MariaDB4j ${MariaDB4j_Repo}
yes | cp -fv $(brew --prefix)/opt/mariadb@$MariaDB4j_MAJORMINOR/bin/* ${MariaDB4j_Repo}/DBs/mariaDB4j-db-mac64-$MariaDB4j_VERSION/src/main/resources/ch/vorburger/mariadb4j/mariadb-$MariaDB4j_VERSION/osx/bin/
yes | cp -fv $(brew --prefix)/opt/mariadb@$MariaDB4j_MAJORMINOR/scripts/* ${MariaDB4j_Repo}/DBs/mariaDB4j-db-mac64-$MariaDB4j_VERSION/src/main/resources/ch/vorburger/mariadb4j/mariadb-$MariaDB4j_VERSION/osx/bin/
mvn install -f ${MariaDB4j_Repo}/DBs/mariaDB4j-db-mac64-$MariaDB4j_VERSION/pom.xml
```

While this does work. fine for local. Be aware that when MariaDB4j makes a next release if the version of MariaDB changes you'll need to run these commands again for the new version.

**Comment by abrousseau001** [comment]
@vorburger Are you still interested in doing the least PITA option of building a GitHub repository that uses GitHub's Actions to build the macOS version of MariaDB? I was just playing around with the idea locally on my M2 MacBook Pro and it seems pretty straight forward for building (at least the version without all the extras included). I am willing to put up a test repository with packages from the action for both intel and arm64 (if the GitHub Action supports both containers).

**Comment by vorburger** [comment]
@abrousseau001 how do you mean building a GitHub repository? As in another repo/project? I am happy to review Pull Requests for THIS repo (https://github.com/vorburger/MariaDB4j) with proposed contributions to make this possible... 😄

**Comment by vorburger** [comment]
BTW as mentioned on #713, https://github.com/ankane/setup-mariadb seems like a great GitHub Action for this project.

It's interesting to note how they use Homebrew to locally install a MariaDB. Perhaps that could be used to "package" it...

**Comment by abrousseau001** [comment]
> @abrousseau001 how do you mean building a GitHub repository? As in another repo/project? I am happy to review Pull Requests for THIS repo (https://github.com/vorburger/MariaDB4j) with proposed contributions to make this possible... 😄

Yes, I meant a Pull Request, which is a fork until you bring it back it. The only contribution to start with would be providing a package for MariaDB. Keeping the possible testing dependency down the road once that is established.

**Comment by abrousseau001** [comment]
> BTW as mentioned on #713, https://github.com/ankane/setup-mariadb seems like a great GitHub Action for this project.
> 
> It's interesting to note how they use Homebrew to locally install a MariaDB. Perhaps that could be used to "package" it...

I think using that action might be a good idea for testing in GitHub actions, but I would think a better approach, if using homebrew, would be to either require homebrew as a dependency to install MariaDB locally, eliminating the above package and building, or getting with the Homebrew folks to see if there is any way to get the pre-built binaries without using the brew command. As a macOS user, I am more than happy to run a single command to get the dependency beforehand, but then you have to look at various scenarios:
- I am unable to access homebrew (network restrictions)
- I downloaded this this Java library to test various versions and have cleanup after the fact
I think ultimately there is a smaller population using MariaDB in macOS, I can be wrong on this, so requiring homebrew and installation beforehand seems like a fine solution to me especially when MariaDB themselves don't provide the package for macOS and tell you to use homebrew.

**Comment by vorburger** [comment]
@abrousseau001 re. _a Pull Request, which is a fork until you bring it back it,_ the trouble is that I still would need a Mac to `mvn deploy` a Mac artifact, no? And I don't have one. Unless I trust somebody (you? or @TheKnowles or so?) to deploy a Mac artifact for MariaDB4j and add a dependency? Or am I totally misunderstanding what you may be suggesting?

@abrousseau001 re. _requiring homebrew and installation beforehand seems like a fine solution to me especially when MariaDB themselves don't provide the package for macOS and tell you to use homebrew_, do you mean Option F. above, with https://github.com/vorburger/MariaDB4j#how-using-existing-native-mariadb-binaries? I personally would actually be fine with leaving it at that, but I think the project is missing at least 2-3 things for that to be "fully done":

1. test coverage for Mac on CI - see #714, initially maybe just "hard-coded"
2. some "automation" along the lines of https://github.com/MariaDB4j/MariaDB4j/issues/699#issuecomment-1465263357, for this Java library to "just to the right thing based on where it's running"; incl. some built-in "guidance" for users; on Mac, that "right thing" might be trying to find a local installation, and if N/A then abort printing something clear and helpful, such as: _"Dude, `brew install mariadb`, as per ... (URL)".

While I personally won't work on this (without Sponsoring), I certainly would review any PRs about this!

PS: If you use MariaDB4j professionally, please consider  **💸 support [on OpenCollective](https://opencollective.com/mariadb4j), via [GitHub Sponsoring](https://github.com/sponsors/vorburger) or through [a Tidelift subscription](https://tidelift.com)** to ensure active maintenance of this project [used by hundreds](https://github.com/MariaDB4j/MariaDB4j/network/dependents), since 2011! 🫶_

**Comment by vorburger** [comment]
I just noticed that https://brew.sh is actually available on Linux as well now, and not just Mac only anymore, and (apparently, not yet actually tested, just read) can "cross" [`brew fetch --os ... --arch ...](https://docs.brew.sh/Manpage) its MariaDB [formula](https://formulae.brew.sh/formula/mariadb)...

... maybe that's a way forward here worth further exploring to allow me to (eventually, once it works) `mvn deploy` this, from Linux.

@abrousseau001 given your comment on #714 last week, are you still interested in creating such a PR? I would definitely welcome the contribution, and be happy to review it.

Testing of a PR for this would require to first have #714.

And if that doesn't work, then running Brew on a Mac GitHub Runner could... but that would then require to be able to directly release from there, which is more work.

**Comment by abrousseau001** [comment]
Give me a bit @vorburger and I will see what I can do on my MacBook Pro to ensure when the request comes in it should work. Ping me again in a couple of weeks if you haven't heard anything new.

**Comment by vorburger** [comment]
@abrousseau001 did you ever get this working?

**Comment by vorburger** [comment]
If anyone is still 😆 reading along here, FYI I'm now toying with this (originally motivated by @Muta-Jonathan [here](https://github.com/MariaDB4j/MariaDB4j/pull/1085#issuecomment-2753469715), and then the renewed discussion with @abrousseau001 over in #1123.

A friend of mine point out that my lack of access to a macOS is "a solvable problem" - by using one of several available 🍎 Apples 🍏 in the ☁ Cloud providers; e.g. https://flow.swiss/mac-bare-metal (hoi zämä [fellow!] 🇨🇭 folks, do you want to sponsor this MariaDB4j project by giving me some credits?).

Can I ask some of you with more experience on this why you were suggesting using `brew install --build-from-source`, above? For me, a `brew install mariadb@11.4` seems to work just fine... and appears to put stuff into  `/opt/homebrew/opt/mariadb@11.4/` which perhaps isn't that difficult for me to (write a script for to) package up and make available - similarly to the Linux and Windows versions.

**Comment by vorburger** [comment]
All right, this wasn't even _that_ difficult, after all... 🎄 #1124 does this! (But it's still WIP, and not available on Maven Central, yet.)

There is just 1 minor hiccup... while that approach won't run `brew` (Homebrew) for users itself (unlike the #1123 approach), it DOES require Homebrew to have been installed, and at least some OTHER tool to previously have installed `openssl`. That's because the Homebrew 🍼 "bottle" 🧴 of MariaDB dynamically links OpenSSL, see:

```sh
$ otool -L DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-11.4.5-macos-arm64/target/generated-resources/ch/vorburger/mariadb4j/mariadb-11.4.5/osx/bin/mariadbd

DBs/mariaDB4j-db-11.4.5/mariaDB4j-db-11.4.5-macos-arm64/target/generated-resources/ch/vorburger/mariadb4j/mariadb-11.4.5/osx/bin/mariadbd:
        /System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1226.0.0)
        /opt/homebrew/opt/pcre2/lib/libpcre2-8.0.dylib (compatibility version 14.0.0, current version 14.0.0)
        /usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.12)
        /opt/homebrew/opt/openssl@3/lib/libssl.3.dylib (compatibility version 3.0.0, current version 3.0.0)
        /opt/homebrew/opt/openssl@3/lib/libcrypto.3.dylib (compatibility version 3.0.0, current version 3.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1351.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1800.105.0)
```

Avoiding this would probably require re-building MariaDB ourselves with static linking. According to [this 🔮 conversation I've just had with Google Gemini about it](https://g.co/gemini/share/3d1cff7cbe6b), that may not be completely trivial. So it's not something I'm particularly motivated to further explore. Unless someone would like to 💰 sponsor me 💸 to figure that out? 🤑 Or if someone else wants to have a go at this, and propose a PR with a solution - go for it! In the meantime, for now, I'll simply document that a `brew install openssl@3` is still a pre-requisite in the README.

**Comment by vorburger** [comment]
> still WIP, and not available on Maven Central, yet

After some 🔏 security related 😮‍💨 "fun" 😷 with GPG and what not in order to be able to do the #1121 equivalent on https://flow.swiss/mac-bare-metal over 🚇 SSH, I have now deployed `ch.vorburger.mariaDB4j:mariaDB4j-db-macos-arm64:11.4.5` to Maven Central! 🥳 It's not yet visible on https://repo.maven.apache.org/maven2/ch/vorburger/mariaDB4j/ (for me), but should be shortly, hopefully.

**Comment by vorburger** [comment]
> It's not yet visible on (...), but should be shortly, hopefully.

https://repo.maven.apache.org/maven2/ch/vorburger/mariaDB4j/mariaDB4j-db-macos-arm64/11.4.5/ - le voilà!

**Comment by vorburger** [comment]
With #1124 merged, I'll consider this (~3.5 years old!) FR as now completed, and am closing this issue.

Please open new issues for anything else related to [macOS](https://en.wikipedia.org/wiki/MacOS) support by MariaDB4j.
